### PR TITLE
[REG2.072a] Issue 15992 - ICE with field variable of instantiated struct

### PR DESCRIPTION
**Pull request by 9rnsr**
When a template is instantiated in function body, it usually gets no chance to run `semantic2`, because inserted module's `semanticRun` would be greater than `PASSsemantic2done` (function bodies are analyzed in global `semantic3` stage).

Therefore we should append template instances to global deferred `semantic2` list as same as for `semantic3`.

---

<del>Essentially this change will increase compilation time. To minimize performance regression, I applied some improvements to `runDeferredSemantic(|2|3)`.</del>

I've removed some followup changes for recovering compilation speed. Read the below discussion.

### CODE REVIEW

**Comment by JackStouffer** [comment]
Can anyone look at this? This is blocking https://github.com/dlang/phobos/pull/4280

**Comment by WalterBright** [comment]
Please don't combine different things into one PR.

**Comment by 9rnsr** [comment]
@WalterBright OK, I dropped all followup optimizations as you request. This PR will surely introduce compilation slow down.

**Comment by WalterBright** [comment]
This is a recent regression - what change caused it? Should we revert that change instead of introducing another slowdown?

**Comment by 9rnsr** [comment]
The new assertion in `todt.d` was introduced by #5700. It had intended all variables would have been completed their `semantic2` already at the glue layer. I had thought the assumption was correct because it has already been achieved by #5500.

But the assumption was actually incorrect because some template instances don't get changes to run their semantic2 yet. I'm trying to fix the issue in this PR.

@WalterBright You can choose the two methods to fix issue 15992:
1. Remove the assertion `assert(vd->semanticRun >= PASSsemantic2done);` in `todt.d`. By that, front-end semantic analysis process will be left broken.
2. Merge this. front-end semantic analysis process will become more correct.

The regression issue is only in git-head (2.072 alpha). So I recommend to choose 2.

**Comment by 9rnsr** [comment]
@WalterBright Maybe you're thinking dmd is already stable, but it's a yet fantasy. We have to continue to advance, otherwise D will die.

**Comment by WalterBright** [comment]
> Maybe you're thinking dmd is already stable

Not at all. But I do not like the long list of regressions, nor do I care for the ever-increasing complexity. There ought to be straightforward, understandable ways of doing things.

**Comment by WalterBright** [comment]
> Essentially this change will increase compilation time.

The semantics all need to be run anyway. Where is the increase?

**Comment by WalterBright** [comment]
@9rnsr, thanks for posting an explanation. This was very helpful.

**Comment by 9rnsr** [comment]
> But I do not like the long list of regressions, nor do I care for the ever-increasing complexity. There ought to be straightforward, understandable ways of doing things.

dmd is still a collection of special cases.

For example, the translation from `search`ed symbol corresponding `Expression` (e.g. `OverloadSet` to `OverExp`) is placed in `TypeStruct.dotExp`, `TypeClass.dotExp`, `DsymbolExp.resolve`, and `DotIdExp.semanticY`. They have different implementations, so we have to determine first which is the correct behavior. #5588, #5676, #5677 are the related PRs about that issue. I'll planning to merge 4 different implementations to DsymbolExp.resolve, but it's blocked by these yet not merged PRs.

In short, we cannot avoid the long regression list today. If dmd is **already** implemented simply and D language spec is precisely defined, the issue fix will also be simple. If you lament about the regression count, you have to update language spec for corner cases and cleanup dmd implementation more and more.

**Comment by 9rnsr** [comment]
> The semantics all need to be run anyway. Where is the increase?

I explain the issue AGAIN: today some template instance members cannot get chances to invoke their `semantic2`. That's a critical problem since D1 or earlier. The new assertion introduced by #5700 just revealed that issue.
After this PR, those semantic2 will be invoked properly. They of course need additional compilation time.

**Comment by JackStouffer** [comment]
@WalterBright can you take a look at this again?

**Comment by WalterBright** [comment]
Auto-merge toggled on
